iroha_actor/src/broker.rs

```diff
@@ -0,0 +1,130 @@
+//! Module with message broker for `iroha_actor`
+
+#![allow(clippy::module_name_repetitions)]
+
+use std::any::{Any, TypeId};
+
+use dashmap::mapref::entry::Entry;
+use dashmap::DashMap;
+use futures::future;
+use once_cell::sync::Lazy;
+
+use super::*;
+
+type TypeMap<V> = DashMap<TypeId, V>;
+type BrokerRecipient = Box<dyn Any + Sync + Send + 'static>;
+
+static BROKER: Lazy<TypeMap<Vec<(TypeId, BrokerRecipient)>>> = Lazy::new(DashMap::new);
+
+/// Trait alias for messages which can be broked
+pub trait BrokerMessage: Message<Result = ()> + Clone + 'static + Send {}
+
+impl<M: Message<Result = ()> + Clone + 'static + Send> BrokerMessage for M {}
+
+fn message_entry<'a>(id: TypeId) -> Entry<'a, TypeId, Vec<(TypeId, BrokerRecipient)>> {
+    BROKER.entry(id)
+}
+
+/// Send message via broker
+pub async fn issue_send<M: BrokerMessage + Send + Sync>(m: M) {
+    let entry = if let Entry::Occupied(entry) = message_entry(TypeId::of::<M>()) {
+        entry
+    } else {
+        return;
+    };
+    let send = entry.get().iter().filter_map(|(_, recipient)| {
+        recipient
+            .downcast_ref::<Recipient<M>>()
+            .map(|recipient| recipient.send(m.clone()))
+    });
+
+    drop(future::join_all(send).await);
+}
```

Comment:
I see that right now broker is just a dynamic dispatch over the global DashMap. This can lead to congestion in the point of getting access to its entries.

Have you considered an alternative design of making Broker a kind of special Actor, with everybody knowing getting its address in their `new` method. So that everybody send broker messages over channel, and then broker sends them to the listener actors also over channels. This will still use dashmap, but now only broker tasks will lock - not the actor handle tasks. What do you think about this design?

Reply:
I think the current design is also quite good. Maybe you can try each of this designs locally, benchmark and in general see which is the best.

Reply:
It won't lead to congestion as we only getting shared reference out of dashmap and not locking. Several shared references of same bucket can be given.

For me seems that this design will work faster. In actix it is made exactly as you propose

Reply:
Ok, I see, then I guess it is ok :smile:

Reply:
I am a bit concerned by once_cell though, what does it use internally?

Reply:
https://docs.rs/once_cell/1.7.2/src/once_cell/imp_pl.rs.html#12-16

Yes, mutex and atomic for signaling that it is inited

Reply:
Okay, I review their impl and it seems good enough. There are two things I would suggest:
1. After we migrate to tokio use https://docs.rs/tokio/1.5.0/tokio/sync/struct.OnceCell.html (but we also have to review their code)
2. Introduce a `broker::init()` method and use once cell instead of lazy cell, this way we can call it at Iroha::start and ensure that it is before the actual actor framework start therefore.

Reply:
Why is it bad to be inited Lazy if it was async? Can we create it by ourselves for tokio (it is just OnceCell and function)?

Reply:
I guess we can force its creation:
https://docs.rs/once_cell/1.7.2/once_cell/sync/struct.Lazy.html#method.force

Reply:
It still would be needed to force it in `broker::init` but yes, we can use it

Reply:
I mean then semantically it will just be OnceCell